File: ielts-listening-ai-trainer/AppFiles/backend/app/services/scoring.py
import re


def normalize(value: str) -> str:
    return re.sub(r"\s+", " ", re.sub(r"[^\w\s'-]", " ", value.lower())).strip()


def levenshtein(a: str, b: str) -> int:
    matrix = [[0] * (len(b) + 1) for _ in range(len(a) + 1)]
    for i in range(len(a) + 1):
        matrix[i][0] = i
    for j in range(len(b) + 1):
        matrix[0][j] = j
    for i in range(1, len(a) + 1):
        for j in range(1, len(b) + 1):
            matrix[i][j] = min(
                matrix[i - 1][j] + 1,
                matrix[i][j - 1] + 1,
                matrix[i - 1][j - 1] + (0 if a[i - 1] == b[j - 1] else 1),
            )
    return matrix[-1][-1]


def words(value: str) -> list[str]:
    return [word for word in normalize(value).split(" ") if word]


def band_estimate(score: int) -> int:
    if score >= 90:
        return 9
    if score >= 82:
        return 8
    if score >= 74:
        return 7
    if score >= 64:
        return 6
    if score >= 52:
        return 5
    if score >= 40:
        return 4
    if score >= 28:
        return 3
    if score >= 16:
        return 2
    return 1


def score_answer(reference: str, answer: str):
    clean_reference = normalize(reference)
    clean_answer = normalize(answer)
    if not clean_answer:
        return {"score": 0, "mistakes": ["vocabulary"]}

    distance = levenshtein(clean_reference, clean_answer)
    longest = max(len(clean_reference), len(clean_answer), 1)
    similarity = max(0, 1 - distance / longest)

    target_words = [word for word in words(reference) if len(word) > 3]
    answer_words = set(words(answer))
    matched = len([word for word in target_words if word in answer_words])
    keyword_match = matched / len(target_words) if target_words else similarity
    score = round(max(0, min(100, similarity * 70 + keyword_match * 30)))

    mistakes = set()
    if re.search(r"\d", reference) and not re.search(r"\d", answer):
        mistakes.add("numbers")
    if score < 84 and similarity > keyword_match:
        mistakes.add("vocabulary")
    if score < 90 and abs(len(words(reference)) - len(words(answer))) > 2:
        mistakes.add("grammar")
    if score < 92 and distance > 0:
        mistakes.add("spelling")

    return {"score": score, "mistakes": sorted(mistakes)}

File: IELTS-Listening-AI-Trainer-v1.1.0/AppFiles/frontend/lib/scoring.ts
export type ScoreResult = {
  score: number;
  mistakes: string[];
};

const normalize = (value: string) =>
  value
    .toLowerCase()
    .replace(/[^\w\s'-]/g, " ")
    .replace(/\s+/g, " ")
    .trim();

const words = (value: string) => normalize(value).split(" ").filter(Boolean);

export function levenshtein(a: string, b: string) {
  const matrix = Array.from({ length: a.length + 1 }, (_, i) => [i]);
  for (let j = 1; j <= b.length; j += 1) matrix[0][j] = j;
  for (let i = 1; i <= a.length; i += 1) {
    for (let j = 1; j <= b.length; j += 1) {
      matrix[i][j] = Math.min(
        matrix[i - 1][j] + 1,
        matrix[i][j - 1] + 1,
        matrix[i - 1][j - 1] + (a[i - 1] === b[j - 1] ? 0 : 1)
      );
    }
  }
  return matrix[a.length][b.length];
}

export function bandEstimate(score: number) {
  if (score >= 90) return 9;
  if (score >= 82) return 8;
  if (score >= 74) return 7;
  if (score >= 64) return 6;
  if (score >= 52) return 5;
  if (score >= 40) return 4;
  if (score >= 28) return 3;
  if (score >= 16) return 2;
  return 1;
}

export function scoreAnswer(target: string, answer: string): ScoreResult {
  const cleanTarget = normalize(target);
  const cleanAnswer = normalize(answer);
  if (!cleanAnswer) {
    return { score: 0, mistakes: ["vocabulary"] };
  }

  const distance = levenshtein(cleanTarget, cleanAnswer);
  const longest = Math.max(cleanTarget.length, cleanAnswer.length, 1);
  const similarity = Math.max(0, 1 - distance / longest);

  const targetWords = words(target).filter((word) => word.length > 3);
  const answerWords = new Set(words(answer));
  const matched = targetWords.filter((word) => answerWords.has(word)).length;
  const keywordMatch = targetWords.length ? matched / targetWords.length : similarity;
  const score = Math.round(Math.min(100, Math.max(0, similarity * 70 + keywordMatch * 30)));

  const mistakes = new Set<string>();
  if (/\d/.test(target) && !/\d/.test(answer)) mistakes.add("numbers");
  if (score < 84 && similarity > keywordMatch) mistakes.add("vocabulary");
  if (score < 90 && Math.abs(words(target).length - words(answer).length) > 2) mistakes.add("grammar");
  if (score < 92 && distance > 0) mistakes.add("spelling");

  return { score, mistakes: Array.from(mistakes) };
}
